src/lib/cloud-manager-hoc.jsx

```diff
@@ -21,55 +21,61 @@ const cloudManagerHOC = function (WrappedComponent) {
             this.cloudProvider = null;
         }
         componentDidMount () {
-            if (this.canUseCloud(this.props)) {
+            if (this.shouldConnect(this.props)) {
                 this.connectToCloud();
             }
         }
         componentDidUpdate (prevProps) {
-            // TODO add disconnection logic when loading a new project
+            // TODO need to add cloud provider disconnection logic and cloud data clearing logic
+            // when loading a new project e.g. via file upload
             // (and eventually move it out of the vm.clear function)
-            // e.g. was previously showingWithId but no longer, but we need
-            // to check that we don't disconnect when saving a project to the
-            // server.
 
-            // If we couldn't use cloud data before, but now we can, try opening a cloud connection
-            if (this.canUseCloud(this.props) && !this.canUseCloud(prevProps)) {
+            if (this.shouldConnect(this.props) && !this.shouldConnect(prevProps)) {
                 this.connectToCloud();
-                return;
             }
 
-            // TODO add scratcher check
-
-            if ((this.props.username !== prevProps.username) ||
-                (this.props.projectId !== prevProps.projectId)) {
-                this.connectToCloud();
+            if (this.shouldDisconnect(this.props) && !this.shouldDisconnect(prevProps)) {
+                this.disconnectFromCloud();
             }
         }
         componentWillUnmount () {
-            if (this.cloudProvider) {
-                this.cloudProvider.requestCloseConnection();
-                this.cloudProvider = null;
-            }
+            this.disconnectFromCloud();
```

Comment:
I think we do want to disconnect when unmounting, and conversely, attempt to connect when mounting. This is so that the disconnect/connect is tightly coupled with the lifecycle of HOC, and we aren't leaving behind unwanted websocket connections when switching between editor mode and player mode etc.

Reply:
So just to be crystal clear, say you have a loop that increments a cloud var

and you run the program, and switch back and forth between project page and editor

and meanwhile, on another account on another browser, you run the same project

what should you see? the cloud var stutter for a second or two on each transition? But it should more or less keep up, right?

I guess that makes sense!